packages/grid/_modules_/grid/hooks/features/useColumnResize.tsx

```diff
@@ -1,204 +1,130 @@
     }
-  }, [headerHeight, logger, stopResize, updateWidth]);
 
-  const handleMouseMove = React.useCallback(
-    (ev: MouseEvent): void => {
-      if (isResizing.current) {
-        const target = ev.currentTarget! as HTMLDivElement;
-        const rect = target.getBoundingClientRect();
+    // Avoid text selection
+    event.preventDefault();
 
-        resizingMouseMove.current = { x: ev.clientX - rect.left, y: ev.clientY - rect.top };
+    colElementRef.current = event.currentTarget.closest('.MuiDataGrid-colCell') as HTMLDivElement;
+    const field = colElementRef.current.getAttribute('data-field') as string;
+    const colDef = apiRef.current.getColumnFromField(field);
 
-        const offsetLeft = !isLastColumn.current ? rect.left : scrollOffset.current * -1;
+    logger.debug(`Start Resize on col ${colDef.field}`);
+    apiRef.current.publishEvent(COL_RESIZE_START);
 
-        let newWidth = ev.clientX - offsetLeft - currentColPosition.current!;
-        newWidth = newWidth > MIN_COL_WIDTH ? newWidth : MIN_COL_WIDTH;
-        updateWidth(newWidth);
-      }
-    },
-    [updateWidth],
-  );
+    colDefRef.current = colDef;
+    colElementRef.current = columnsRef.current!.querySelector(
+      `[data-field="${colDef.field}"]`,
+    ) as HTMLDivElement;
 
-  // This a hack due to the limitation of react as I cannot put columnsRef in the dependency array of the effect adding the Event listener
-  const columnsRefState = useStateRef(columnsRef);
+    colCellElementsRef.current = findCellElementsFromCol(colElementRef.current) as NodeListOf<
+      Element
+    >;
 
-  React.useEffect(() => {
-    if (columnsRef && columnsRef.current) {
-      logger.info('Adding resizing event listener');
-      const columnsRefEvents = columnsRef.current;
-      columnsRef.current.addEventListener('mouseup', stopResize);
-      columnsRef.current.addEventListener('mouseleave', handleMouseLeave);
-      columnsRef.current.addEventListener('mouseenter', handleMouseEnter);
-      columnsRef.current.addEventListener('mousemove', handleMouseMove);
-
-      return () => {
-        columnsRefEvents.removeEventListener('mouseup', stopResize);
-        columnsRefEvents.removeEventListener('mouseleave', handleMouseLeave);
-        columnsRefEvents.removeEventListener('mouseenter', handleMouseEnter);
-        columnsRefEvents.removeEventListener('mousemove', handleMouseMove);
-      };
-    }
-    return undefined;
-  }, [
-    columnsRefState,
-    columnsRef,
-    handleMouseLeave,
-    handleMouseDown,
-    handleMouseMove,
-    handleMouseEnter,
-    logger,
-    stopResize,
-  ]);
+    const doc = ownerDocument(apiRef.current.rootElementRef!.current as HTMLElement);
+    doc.body.style.cursor = 'col-resize';
+
+    initialOffset.current =
+      (colDefRef.current.width as number) -
+      (event.clientX - colElementRef.current!.getBoundingClientRect().left);
+
+    // @ts-expect-error fixed in v5
+    doc.addEventListener('mousemove', handleResizeMouseMove);
```

Comment:
again why do we need this `// @ts-expect-error` => bad practice. Errors have to be fixed

Reply:
the error is fixed in v5, we need the expect error so the CI fails once we upgrade (so we don't forget to remove the comment).

Reply:
I wouldn't use it and it's broken
What does it bring? Just use a ref...

Reply:
Only the type is wronk, the implementation is working correctly. The problem the hook solved is explain in https://github.com/facebook/react/issues/14099#issuecomment-440013892.

Reply:
It's a small fn with wrong types. => I would not use it, you can rewrite it properly 
```
const methodRef = useRef(method)

useEffect(()=> {
 methodRef.current =method;
}, [method])

const wrapMethod = useCallback((...args)=> methodRef.current(args), []);
return wrapMethod;

```

Reply:
> I would not use it, you can rewrite it properly

If you want to contribute the change to the core, that would be great! Note that we need to use the layout effect, so `useEnhancedEffect`. Regarding `(0, ref.current)` it's meant to pass along the event, but we can probably drop it. I have never used it, and I doubt people importing the module rely on this property.

Reply:
I guess the annoying part with the `// @ts-expect-error` is that it breaks the storybook compilation if you start it and then edit some file (at least that was the case for me).

Reply:
> I guess the annoying part with the // @ts-expect-error is that it breaks the storybook compilation if you start it and then edit some file (at least that was the case for me).

Interesting, how do you reproduce?

Two thoughts.

1. I have never used the Storybook so far for working on a problem. I think that it's only relevant for cases where:

- You can't cover it in the documentation
- You can't write an automated test for it

I haven't been working on a feature that matches the above yet. But I didn't build much so, that's kind of expected.

2. I personally think that a TypeScript issue shouldn't fail the compilation in development (only in production). But that's a personal dev preference. I think that we should make it configurable. On the docs, we have explicitly disabled the behavior: https://github.com/mui-org/material-ui/pull/20128.